src/core/qgspointlocator.cpp

```diff
@@ -574,27 +586,107 @@ void QgsPointLocator::setRenderContext( const QgsRenderContext *context )
 
 }
 
-bool QgsPointLocator::init( int maxFeaturesToIndex )
+void QgsPointLocator::onInitTaskTerminated()
 {
-  return hasIndex() ? true : rebuildIndex( maxFeaturesToIndex );
+  mIsIndexing = false;
+  mRenderer.reset();
+  mSource.reset();
 }
 
+void QgsPointLocator::onRebuildIndexFinished( bool ok )
+{
+  onInitTaskTerminated();
+
+  // treat added and deleted feature while indexing
+  for ( QgsFeatureId fid : mAddedFeatures )
+    onFeatureAdded( fid );
+
+  for ( QgsFeatureId fid : mDeletedFeatures )
+    onFeatureDeleted( fid );
+
+  emit initFinished( ok );
+}
+
+bool QgsPointLocator::init( int maxFeaturesToIndex, bool relaxed )
+{
+  const QgsWkbTypes::GeometryType geomType = mLayer->geometryType();
+  if ( geomType == QgsWkbTypes::NullGeometry // nothing to index
+       || hasIndex()
+       || mIsIndexing ) // already indexing, return!
+    return true;
+
+  mRenderer.reset( mLayer->renderer() ? mLayer->renderer()->clone() : nullptr );
+  mSource.reset( new QgsVectorLayerFeatureSource( mLayer ) );
+
+  if ( mContext )
+  {
+    mContext->expressionContext() << QgsExpressionContextUtils::layerScope( mLayer );
+  }
+
+  mIsIndexing = true;
+
+  if ( relaxed )
+  {
+    mInitTask = new QgsPointLocatorInitTask( this );
+    connect( mInitTask, &QgsPointLocatorInitTask::rebuildIndexFinished, this, &QgsPointLocator::onRebuildIndexFinished );
+    connect( mInitTask, &QgsPointLocatorInitTask::taskTerminated, this, &QgsPointLocator::onInitTaskTerminated );
+    QgsApplication::taskManager()->addTask( mInitTask );
+    return true;
+  }
+  else
+  {
+    const bool ok = rebuildIndex( maxFeaturesToIndex );
+    mIsIndexing = false;
+    emit initFinished( ok );
+    return ok;
+  }
+}
+
+void QgsPointLocator::waitForFinished() const
+{
+  mInitTask->waitForFinished();
+
+  // force process of signal emitted from task thread
+  // so rebuildIndexFinished and taskTerminated are called
+  QCoreApplication::processEvents();
```

Comment:
Hmm... whenever I see processEvents() call, I get really scared :-) This call is very unpredictable in its side effects unfortunately... for example, it can also get into processing of the next mouse move event (while in the middle of processing the current mouse move event) and completely break the expected flow of code, often leading to crashes.

Is there any chance we can get rid of sending a signal from the task from the background thread and instead just use the services of task manager to find out that the task has finished?

Reply:
Yes, this way should be better d246d9f